README.md
# Checkers-Game
A GUI checkers game implementation in Intel Assembly

The game includes a dou mode, in which two players compete each-other in a turn-based game, and a single mode, in which the player competes against an AI agent.

This project was made as a final project for the Height's program first year.

Made for educational purposes.


changeLog.txt
buckup_1: only board printing
buckup_2: players and circles added
buckup_3: changed "test.asm" to "main.asm".
	  added selecting panel on click and painting it
	  added nextTurn procedure
buckup_4: mark possible moves and more base procedures
buckup_5: mark two sides, deselect and exit on esc
buckup_6: movement and eating
buckup_7: Double eating, still have problems (see TODO)
buckup_8: Fixed double eating bugs.
buckup_9: added two base functions for AI, started menu.
buckup_10: Menu finished (enter for selecting)
	   Added the function getDefensiveScore for the AI, and finished other AI procedures.
buckup_11: finished the getScore AI proc, and the getOffensiveScore proc, together with helping procedures.
buckup_12: Fixed moving and eating bugs, started AITurn
buckup_13: AI is moving
buckup_14: AI is eating
buckup_15: PBEMoveToBlockEating: can eat as respond, and move others to block eating
buckup_16: now can move away by it self
buckup_17: game over recognition and screen
buckup_18: fixed offensiveScore bug
buckup_19: re-wrote the 'getBeingEatenTimes' proc from scratch.
	   fixed getDScore by adding the moving simulation
	   fixed 'IsExposedForEatingSide' by putting in bx the 'to' panel adress before the end
buckup_20: final
buckup_21: fixed moving to block being eaten
buckup_22: fixed double eating still selected at end bug
buckup_23: fixed double eating of nothing
buckup_24: now asking if the player is sure he wants to exit in game-mode when clicks Esc


Code/menu.asm
; startMenu PROCEDURE HEADER
; 	parameters: none
; 	return: void

proc startMenu
  push ax
  push bx
  push cx
  push dx

  call renderMenu
  xor bx, bx ; will hold which option is currently on (0 for player, 1 for AI)
  SMInputListening:

    ; bring the cursor again
    mov ax, 1
    int 33h

    call getMouseAction

    ; hide the cursor, so we will not get the cursor color
      mov ax, 2
      int 33h

    cmp cx, -1 ; if enter was clicked
      je SMAfterInput

    ; get the color of the clicked pixel
    mov ah, 0Dh
    mov bh, 0
    int 10h ; now, al contains the color

    cmp al, RMArrowColor
     jne dSMDontSwitchOptions

     cmp cx, 100
      jbe SMChaneToPlayerGM
      cmp bx, 1
        je SMInputListening ; if AI is already the shown option, do not re-render
        inc bx ; make bx 1
      jmp SMRnderGameMode
    SMChaneToPlayerGM:
      cmp bx, 0
        je SMInputListening ; if player is already the shown option, do not re-render
      dec bx
    SMRnderGameMode:
      push bx ; the option
      call renderAsGameMode

   dSMDontSwitchOptions:

  jmp SMInputListening
  SMAfterInput:

    ; bring the cursor again
    mov ax, 1
    int 33h

    mov byte ptr [PvWhat], bl

  pop dx
  pop cx
  pop bx
  pop ax
  ret
endp startMenu

; renderMenu PROCEDURE HEADER
; 	parameters: none
; 	return: void

RMMarginTop       equ 20
RMMarginLeft      equ 53
RMBlockSize       equ 3
RMMaxNumOfLetters equ 8
RMArrowColor     equ 44

proc renderMenu
  push dx
  push bx
  push ax
  push si

  ; hide the cursor before printing: will not print on cursor location otherwise
    mov ax, 2
    int 33h

  mov ax, 0
  RMHrLoop:

    push ax
    push RMMarginTop + 5*3 + 3*3
    mov bx, 1
    and bx, ax
    cmp bx, 1
      je RMRed
      push 1
      jmp RMCallDrawCircle
    RMRed:
      push 4

    RMCallDrawCircle:
    call drawCircle

  add ax, 21
  cmp ax, 310
  jb RMHrLoop

  ; printing the menu
    ; render the title
      push 15
      push offset titleLine1
      push RMMarginLeft
      push RMMarginTop
        call renderWord

    ; render the PLAYER word
      push 15
      push offset playerLine1
      push RMMarginLeft
      push RMMarginTop + 60
        call renderWord

    ; render the VS word
      push 15
      push offset vsLine1
      push RMMarginLeft
      push RMMarginTop + 60 + 5*3 + 3
        call renderWord

    ; arrows
      push RMArrowColor
      push 1
      push RMMarginLeft + 3
      push RMMarginTop + 60 + 5*3 + 3 + 46
        call drawTriangle

      push RMArrowColor
      push 0
      push RMMarginLeft + 8*3*8 + 5
      push RMMarginTop + 60 + 5*3 + 3 + 46
        call drawTriangle

    ; render options box
      push 0
      call renderAsGameMode

    ; bring the cursor again
      mov ax, 1
      int 33h

  pop si
  pop ax
  pop bx
  pop dx
  ret
endp renderMenu

; renderWord PROCEDURE HEADER
; 	parameters: color, offset, x, y
; 	return: void

  RWColor equ  [bp+10]
  RWOffset equ  [bp+8]
  RWX      equ  [bp+6]
  RWY      equ  [bp+4]

proc renderWord
  push bp
  mov bp, sp
  push dx
  push bx
  push ax
  push si

  mov si, RWOffset
  mov ax, RWY ; the y
  mov dx, RWX ; the x

  RMPrintingLoop:

    mov bl, [si]
    add dx, RMBlockSize * 8 ; 53+3*8

    mov cx, 8
    RMBitLoop:

      shr bl, 1
      jnc RMDontRender
        push RMBlockSize
        push RMBlockSize
        push dx
        push ax
        push RWColor ; white
        call drawSquare
      RMDontRender:
      sub dx, RMBlockSize

    loop RMBitLoop

    add dx, 8 * RMBlockSize + 2

    mov bx, RWX
      add bx, 8 * ( RMMaxNumOfLetters * RMBlockSize) ; 53 + 8*8*3 = 245
    cmp dx, bx
      jbe RMNotNextLine

      mov dx, RWX
      add ax, RMBlockSize

      RMNotNextLine:

    inc si

    mov bx, RWOffset
      add bx, 5 * RMMaxNumOfLetters
    cmp si, bx
      jb RMPrintingLoop


  pop si
  pop ax
  pop bx
  pop dx
  pop bp
  ret 8
endp renderWord

; renderAsGameMode PROCEDURE HEADER
; 	parameters: the option (0 for player, 1 for AI)
; 	return: void

RAGMOption equ [bp+4]
optionsY equ 139
optionsX equ 72

proc renderAsGameMode
  push bp
  mov bp, sp
  push ax

  push 34
  push 8*8*3+7-19-6
  push optionsX
  push optionsY
  push 2; 48
    call drawSquare

  mov ax, offset playerTallLine1
  cmp word ptr RAGMOption, 1
    jne RAGMDonNotChangeToPvP
    mov ax, offset AILine1

  RAGMDonNotChangeToPvP:

  push 0
  push ax
  push RMMarginLeft
  push optionsY + 2
    call renderWord

  add ax, 5*8
  push 0
  push ax
  push RMMarginLeft
  push optionsY + 2 + 5*3
    call renderWord

  pop ax
  pop bp
  ret 2
endp renderAsGameMode

; exitGameMenu PROCEDURE HEADER
; 	parameters: none
; 	return: void
;
; set position:
; mov ah, 2
; mov bh, 0
; mov dh, 10
; mov dl, (80-35)/2
; int 10h

  proc exitGameMenu

    ; border
      push 80
      push 160
      push (320-160)/2
      push (200-80)/2
      push 3
      call drawSquare

    ; inner color
      push 80-6
      push 160-6
      push (320-160)/2 + 3
      push (200-80)/2 + 3
      push 0
      call drawSquare

    ; write the title
      push 15
      push offset exitTitle1
      push RMMarginLeft
      push 70
      call renderWord

    ; write the options
      push 15
      push offset exitOption1
      push RMMarginLeft
      push 110
      call renderWord

    mov EGMCurrOption, 0
    push 2
    push 5*3 * 2
    push RMMarginLeft + 8*3*5 + 5*2
    push 110 + 5*3 + 4
    push 11
    call drawSquare

    EGMLoop: ; at D90
      mov ah, 1
      int 16h
        jz EGMLoop ; if no button clicked, wait for action
      ; get button clicked
        mov ah, 0
        int 16h
        cmp al, 121 ; if Y
          je EGMChangeOption
        cmp al, 110 ; if n
          je EGMChangeOption
        cmp al, 0Dh ; if enter
          je EGMSelectedOption
      jmp EGMLoop

      EGMChangeOption: ; will change to the one in al
        ; delete the orev underscore
          push 2
          push 160-6
          push (320-160)/2 + 3
          push 110 + 5*3 + 4
          push 0
          call drawSquare

          jmp GMEAfterUglyPieceOfCode

        ; I know its ugly, but I have to stuck this piece of code here. otherwise, it will be out of jumping range...
                      EGMSelectedOption:
                      cmp EGMCurrOption, 1
                        jne EGMFinish
                      call finishProgram
        ; end of that

        GMEAfterUglyPieceOfCode:

        mov EGMCurrOption, 0 ; move 'no' to the selection as default
        push 2 ; push the underscore height
        push 5*3 * 2 ; push the underscore width

        ; now, will push the x according to the option. if y, move 'yes' to the selection
        cmp al, 121 ; if y
          jne EGMChangeToNo
          mov EGMCurrOption, 1
          push RMMarginLeft + 8*3*2 + 2*2
          jmp EGMAgterSettingX
        EGMChangeToNo:
          push RMMarginLeft + 8*3*5 + 5*2

        EGMAgterSettingX:
        push 110 + 5*3 + 4
        push 11
        call drawSquare

        jmp EGMLoop

      EGMFinish:

      call renderBoard

    ret
  endp exitGameMenu


Code/logic.asm
; nextTurn PROCEDURE HEADER
; 	parameters: none
; 	return: void

	proc nextTurn
		push ax
		push bx
		push cx
		push dx

			NTGetMouseAgain:
				call deselect
				call getMouseAction

			; if the x is out of board, get mouse again
			cmp cx, marginLeft
				jb NTGetMouseAgain
			cmp cx, marginLeft + (panelSize * 8)
				ja NTGetMouseAgain

			; now, getting the dataseg location using two procedures
			push cx ; pixels x
			push dx ; pixels y
			call pixelLocationToDatasegLocation ; will return the dataseg location of the clicked panel, and store it in bx

			; checking if the panel is legal for selecting
				push bx
				call isLegalForSelection
				cmp ax, false
				je NTGetMouseAgain ; if the panel selected is not legal, get the mouse action again

			mov word ptr[selectedPanel], bx ; put in selectedPanel the selected pane dataseg address

			call markMovingPossiblities

			; making the panel selected, by turning its third bit on (look at the dataseg details)
			or byte ptr [bx], 1000b
				; render the changed panel
				push bx
				call renderPanel

			; move
			NTMovementMouseLoop:
				call getMouseAction
				; get the dataseg location
					push cx ; pixels x
					push dx ; pixels y
					call pixelLocationToDatasegLocation ; will return the dataseg location of the clicked panel, and store it in bx

					; check if its the same panel as selected, and if it is,
					; get the mouse again (without deselecting). needed to prevent double respond to a single click
						cmp bx, word ptr[selectedPanel]
							je NTMovementMouseLoop
					; if not the selected panel, check if a marked
						push bx
						call isMarkedPanel ; return in ax. if not marked, return false. if yes, return true
						cmp ax, false ; if not marked, deselect and get mouse selection again
							je NTGetMouseAgain

				; if arrived here, selected a marked panel
				NTLegalMovementPanel:
					push bx ; saving for later the panel to which we should move to (move proc)
          push bx ; saving for recursive eating, to know where the player is after moving
					push bx
						call getEatenPanel
						cmp bx, false ; if no player is eaten in the moving action, move and finish
							je NTMoveAndFinish

						NTPlayerWasEating:
							and byte ptr [bx], 01b ; remove the player from the eaten panel
							push bx
							call renderPanel

							call move
							call deselect

							pop bx
							mov word ptr[selectedPanel], bx
								call eatRecursive

							jmp NTFinish

			NTMoveAndFinish:

			call move ; the location was pushed before the eating checks
			pop bx

			NTFinish:

			pop dx
			pop cx
			pop bx
			pop ax
		ret
	endp nextTurn

	; eatRecursive PROCEDURE HEADER
	; 	parameters: none
	; 	return: void

	proc eatRecursive
		push ax

		mov bx, [selectedPanel]
		call markMovingPossiblities
		xor al, al ; to check if two sides have no eating possibility. ax will be increased each time I deselect, and if
							 ; its value is 2 after checkings, both directions have no eating possibility.

		; check if there is an eating possibility, and where there isn't, deselect
			cmp word ptr [EatingRightAdress], 0
				je ERDeselectRight
			ERCheckLeftEatingPoss:
			cmp word ptr [EatingLeftAdress], 0
				je ERDeselecLeft
			cmp byte ptr [recursiveEatingNoMove], 1 ; if no movement option, don't act
				je ERFinish

		jmp DSAfterDeselecting

		ERDeselectRight:
			push 0
			call deselectSide
			inc ax
			jmp ERCheckLeftEatingPoss
		ERDeselecLeft:
			push 1
			call deselectSide
			inc ax

		DSAfterDeselecting:
			; first, making sure that there is a eating possibility (why comparing ax to 2? check out the forth line in this procedure)
			cmp ax, 2
				je ERFinish
      ; making the panel selected, by turning its fourth bit on (look at the dataseg details)
      or byte ptr [bx], 1000b
        ; render the changed panel
        push bx
        call renderPanel

      ; getting mouse input
      ; from here on, very similar to nextTurn action
      ERMouseLoop:
      call getMouseAction
        ; get the dataseg location
          push cx ; pixels x
          push dx ; pixels y
          call pixelLocationToDatasegLocation ; will return the dataseg location of the clicked panel, and store it in bx

        ; check if its the same panel as selected, and if it is,
        ; get the mouse again (without deselecting). needed to prevent double respond to a single click
          cmp bx, word ptr[selectedPanel]
            je ERMouseLoop

        ; if not the selected panel, check if a marked
          push bx
          call isMarkedPanel ; return in ax. if not marked, return false. if yes, return true
          cmp ax, false ; if not marked, end the turn
            je ERFinish

        ; eating
          push bx ; saving for later the panel to which we should move to (move proc)
          push bx ; saving for recursive eating, to know where the player is after moving
          push bx
						call getEatenPanel ; now, the player we ate dataseg address in bx

          and byte ptr [bx], 01b ; remove the player from the eaten panel
          push bx
          call renderPanel

          call move
					call deselect

        ; call the eatRecursive proc again
          pop bx
          mov word ptr[selectedPanel], bx
						call deselect
            call eatRecursive

		ERFinish:

			call deselect

		pop ax
		ret
	endp eatRecursive

;===============================================================================
;																	HELPERS
;===============================================================================


; getEatenPanel PROCEDURE HEADER
; 	parameters: the moving-to panel adress
; 	return: the eaten panel address in bx, or false if not eaten

	GEPMovingToPanel equ [bp+4]

	proc getEatenPanel
		push bp
		mov bp, sp

		mov bx, GEPMovingToPanel

		; first, lets check if the player moved right or left
		cmp bx, [markedPanelRight]
			je GEPRightEatingCheck
			jmp GEPLeftEatingCheck

		; now, lets check if the movement is eating someone
			GEPRightEatingCheck:
				mov bx, word ptr [EatingRightAdress] ; keep the address of the eaten panel
				cmp bx, 0
					jne GEPPlayerWasEating
			jmp GEPReturnFalse

			GEPLeftEatingCheck:
				mov bx, word ptr [EatingLeftAdress] ; keep the address of the eaten panel
				cmp bx, 0
					jne GEPPlayerWasEating

			GEPReturnFalse:
				mov bx, false

			GEPPlayerWasEating:

		pop bp
		ret 2
	endp getEatenPanel

; move PROCEDURE HEADER
; 	parameters: dataseg address of the clicked panel, from where in selectedPanel
; 	return: void

	MDatasegAddress equ [bp+4]

	proc move
		push bp
		mov bp, sp
		push bx

		; check the player turn, and add player in the player turn color to the clicked panel
			mov bx, MDatasegAddress
			or byte ptr[bx], 10b
		; if player 2, change the panel player color
			push [selectedPanel]
			call getPlayer
			cmp ax, PLAYER_2_CODE
			jne MDontChangePlayerColor
			or byte ptr[bx], 100b

		MDontChangePlayerColor:
		push bx
		call renderPanel

		; deselect
			call deselect
		; remove player from the prev panel, and render it
			mov bx, [selectedPanel]
			and byte ptr[bx], 01b
			push bx
			call renderPanel

		pop bx
		pop bp
		ret 2
	endp move


; isMarkedPanel PROCEDURE HEADER
; 	parameters: dataseg address
; 	return: true / false in ax. Whether this panel is marked

	IMPDatasegAddress equ [bp+4]

	proc isMarkedPanel
		push bp
		mov bp, sp
		push bx

		mov ax, true
		mov bx, IMPDatasegAddress
		cmp word ptr[markedPanelRight], bx
			je IMPAfterCheck
		cmp word ptr[markedPanelLeft], bx
			je IMPAfterCheck
		mov ax, false

		IMPAfterCheck:

		pop bx
		pop bp
		ret 2
	endp isMarkedPanel

; getMouseAction PROCEDURE HEADER
; 	parameters: none
; 	return: the mouse click X in cx, and mouse click Y in dx
;		if clicked enter, will return -1 in cx. if clicked esc, will exit the program
	proc getMouseAction
		push bx
		push ax

		GMAWaitingForMouse:
			; check for mouse action, and check if its the left button
				mov ax, 3
				int 33h ; bx contains clicking status
				cmp bx, 1
					je mouseDetected ; keep waiting as long as the left mouse button wasn't clicked
			; check if Esc was pressed, and if it did, exit the game
				mov ah, 1
				int 16h
					jz GMAWaitingForMouse ; if no button clicked, wait for action
				; get button clicked
					mov ah, 0
					int 16h
					cmp al, 0Dh ; enter
						je GMAEnterPressed
					cmp al, 27
						jne GMAWaitingForMouse ; if buttonClicked != Esc, wait for action
			; here if Esc clicked. if in menu, quit. otherwise, ask if want to exit.
				cmp byte ptr [gameStage], 1
					je GMANotInMenu
					call finishProgram
				GMANotInMenu:
				call exitGameMenu

		GMAEnterPressed:
			mov cx, -1
			jmp GMAAfterDividing

		mouseDetected:
			shr cx, 1 ; divide cx by 2

		GMAAfterDividing:
		pop ax
		pop bx
		ret
	endp getMouseAction

	;===============================================================================
	;															 PANEL INTERPETATION
	;===============================================================================

	; getPlayer PROCEDURE HEADER
	; 	parameters: the panel dataseg location
	; 	return: the player code in ax (PLAYER_1_CODE, PLAYER_2_CODE, NO_PLAYER_CODE).

	GPDatasegAddress equ [bp+4]

	proc getPlayer
		push bp
		mov bp, sp
		push bx

		mov bx, GPDatasegAddress
		mov ax, [bx]
		shr ax, 1 ; check panel color
		jc notBlackPanel
			jmp GPReturnNoPlayer
		notBlackPanel:
			shr ax, 1 ; check if there if a player there
			jnc GPReturnNoPlayer
			shr ax, 1 ; check the player color
			jc GPReturnPlayer2 ; return accordingly
				mov ax, PLAYER_1_CODE
				jmp GPDone
			GPReturnPlayer2:
				mov ax, PLAYER_2_CODE
				jmp GPDone


		GPReturnNoPlayer:
			mov ax, NO_PLAYER_CODE ; if a black panel, the player can't be there
		GPDone:

		pop bx
		pop bp
		ret 2
	endp getPlayer


	; finishProgram PROCEDURE HEADER
	; 	parameters: none
	; 	return: void

	proc finishProgram
		; switch back to text mode
		mov ah,0
		mov al,2
		int 10h

		exit:
			mov ax, 4c00h
			int 21h

		ret
	endp finishProgram


Code/main.asm
;	NOTE: alt+x+enter for fullscreen in dosbox

MODEL small
STACK 100h
DATASEG

	; !IMPORTANT!: keep the board first in the dataseg
	; table panels value definitons:
	;		00 = black empty
	;		1 = white empty
	;			011 = white with player 1 in it
	;			111 = white with player 2 in it
	;		1xxx = selected (clicked on turn)
	;		10xxx = marked (for movement)
	;		is marked		is selected		Player2/Player1		there is a player? 		panel color
	;			 0/1					0/1							0/1		  				  0/1								  0/1

	row1 db 4 dup (0,    011b)
	row2 db 4 dup (011b, 0)
	row3 db 4 dup (0,    011b)
	row4 db 4 dup (1,    0)
	row5 db 4 dup (0,    1)
	row6 db 4 dup (111b, 0)
	row7 db 4 dup (0,    111b)
	row8 db 4 dup (111b, 0)

	turn db 1

	marginLeft 	equ 60		; (width-height)/2 = (320 - 200) / 2
	panelSize 	equ 25		;	boardSize/panelsAmount = 200/8
	; colors
		white 				db 15
		black 				db 0
		player1Color 	db 4  ; red
		player2Color 	db 1  ; blue
		selectedColor db 13 ; pink
		markedColor		db 14 ; yello

		PLAYER_1_CODE  equ 1
		PLAYER_2_CODE  equ 2
		AI_CODE 			 equ 2
		NO_PLAYER_CODE equ 0

		PLAYER_1_TURN equ 0
		PLAYER_2_TURN equ 1
		true equ 1
		false equ 0

	circle db 7, 11, 13, 15, 17, 19, 19, 7 dup (21), 19, 19, 17, 15, 13, 11, 7

	selectedPanel     dw 0 ; the dataseg address of the selected panel
	markedPanelRight  dw 0
	markedPanelLeft   dw 0
	EatingRightAdress dw 0
	EatingLeftAdress  dw 0

	recursiveEatingNoMove 		db 0 ; will be on if a recursive proc has gone off table, or when trying to eat 2 (good for checking eating possibities)

	; AI
	PWalkFrom dw -1
	PWalkTo		dw -1


	titleLine1 	db 	11111111b, 11100111b, 11111111b, 11111111b, 11000011b, 11111111b, 11111000b, 11111111b
	titleLine2 	db 	11100000b, 11100111b, 11100000b, 11100000b, 11001100b, 11100000b, 11001100b, 11000000b
	titleLine3 	db 	11100000b, 11111111b, 11111111b, 11100000b, 11110000b, 11111111b, 11111000b, 11111111b
	titleLine4 	db	11100000b, 11100111b, 11100000b, 11100000b, 11001100b, 11100000b, 11001100b, 00000011b
	titleLine5 	db	11111111b, 11100111b, 11111111b, 11111111b, 11000011b, 11111111b, 11000111b, 11111111b

	playerLine1 	db  00000000b, 11111110b, 11000000b, 00011000b, 11000011b, 11111111b, 11111000b, 00000000b
	playerLine2 	db	00000000b, 11000011b, 11000000b, 01100110b, 01100110b, 11100000b, 11001100b, 00000000b
	playerLine3 	db	00000000b, 11111110b, 11000000b, 11000011b, 00111100b, 11111111b, 11111000b, 00000000b
	playerLine4 	db	00000000b, 11000000b, 11111111b, 11111111b, 00011000b, 11100000b, 11001100b, 00000000b
	playerLine5 	db	00000000b, 11000000b, 11111111b, 11000011b, 00011000b, 11111111b, 11000111b, 00000000b

	vsLine1	db	00000000b, 00000000b, 00000000b, 11000011b, 11111111b, 00000000b, 00000000b, 00000000b
	vsLine2	db	00000000b, 00000000b, 00000000b, 11000011b, 11000000b, 00000000b, 00000000b, 00000000b
	vsLine3	db	00000000b, 00000000b, 00000000b, 01100110b, 11111111b, 00000000b, 00000000b, 00000000b
	vsLine4	db	00000000b, 00000000b, 00000000b, 00111100b, 00000011b, 00000000b, 00000000b, 00000000b
	vsLine5	db	00000000b, 00000000b, 00000000b, 00011000b, 11111111b, 00000000b, 00000000b, 00000000b

	AILine1  db 00000000b, 00000000b, 00000000b, 00011000b, 11111111b, 00000000b, 00000000b, 00000000b
	AILine2  db 00000000b, 00000000b, 00000000b, 01100110b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine3  db 00000000b, 00000000b, 00000000b, 11000011b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine4  db 00000000b, 00000000b, 00000000b, 11000011b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine5  db 00000000b, 00000000b, 00000000b, 11111111b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine6  db 00000000b, 00000000b, 00000000b, 11000011b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine7  db 00000000b, 00000000b, 00000000b, 11000011b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine8  db 00000000b, 00000000b, 00000000b, 11000011b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine9  db 00000000b, 00000000b, 00000000b, 11000011b, 00011000b, 00000000b, 00000000b, 00000000b
	AILine10 db 00000000b, 00000000b, 00000000b, 11000011b, 11111111b, 00000000b, 00000000b, 00000000b

	playerTallLine1 	db  00000000b, 11111100b, 11000000b, 00011000b, 11000011b, 11111111b, 11111100b, 00000000b
	playerTallLine2 	db	00000000b, 11000110b, 11000000b, 01100110b, 11000011b, 11111111b, 11000110b, 00000000b
	playerTallLine3 	db	00000000b, 11000011b, 11000000b, 11000011b, 01100110b, 11000000b, 11000011b, 00000000b
	playerTallLine4 	db	00000000b, 11000110b, 11000000b, 11000011b, 01100110b, 11000000b, 11000110b, 00000000b
	playerTallLine5 	db	00000000b, 11111100b, 11000000b, 11111111b, 00111100b, 11111111b, 11111100b, 00000000b
	playerTallLine6 	db  00000000b, 11000000b, 11000000b, 11000011b, 00011000b, 11111111b, 11110000b, 00000000b
	playerTallLine7 	db	00000000b, 11000000b, 11000000b, 11000011b, 00011000b, 11000000b, 11011000b, 00000000b
	playerTallLine8 	db	00000000b, 11000000b, 11000000b, 11000011b, 00011000b, 11000000b, 11001100b, 00000000b
	playerTallLine9 	db	00000000b, 11000000b, 11111111b, 11000011b, 00011000b, 11111111b, 11000110b, 00000000b
	playerTallLine10 	db	00000000b, 11000000b, 11111111b, 11000011b, 00011000b, 11111111b, 11000011b, 00000000b

	gameOverTitle1r1 	db  00000000b, 00000000b, 11111110b, 00011000b, 01000010b, 11111111b, 00000000b, 00000000b
	gameOverTitle1r2 	db	00000000b, 00000000b, 11000000b, 01100110b, 11100111b, 11100000b, 00000000b, 00000000b
	gameOverTitle1r3 	db	00000000b, 00000000b, 11001111b, 11000011b, 11011011b, 11111111b, 00000000b, 00000000b
	gameOverTitle1r4 	db	00000000b, 00000000b, 11000110b, 11111111b, 11011011b, 11100000b, 00000000b, 00000000b
	gameOverTitle1r5 	db	00000000b, 00000000b, 11111110b, 11000011b, 11000011b, 11111111b, 00000000b, 00000000b

	gameOverTitle2r1 	db  00000000b, 00000000b, 11111111b, 11000011b, 11111111b, 11111000b, 00000000b, 00000000b
	gameOverTitle2r2 	db	00000000b, 00000000b, 11000011b, 11000011b, 11100000b, 11001100b, 00000000b, 00000000b
	gameOverTitle2r3 	db	00000000b, 00000000b, 11000011b, 01100110b, 11111111b, 11111000b, 00000000b, 00000000b
	gameOverTitle2r4 	db	00000000b, 00000000b, 11000011b, 00111100b, 11100000b, 11001100b, 00000000b, 00000000b
	gameOverTitle2r5  db	00000000b, 00000000b, 11111111b, 00011000b, 11111111b, 11000111b, 00000000b, 00000000b

	exitTitle1 db  00000000b, 00000000b, 11111111b, 11000011b, 01111110b, 11111111b, 00000000b, 00000000b
	exitTitle2 db  00000000b, 00000000b, 11100000b, 01100110b, 00011000b, 00011000b, 00000000b, 00000000b
	exitTitle3 db  00000000b, 00000000b, 11111111b, 00111100b, 00011000b, 00011000b, 00000000b, 00000000b
	exitTitle4 db  00000000b, 00000000b, 11100000b, 01100110b, 00011000b, 00011000b, 00000000b, 00000000b
	exitTitle5 db  00000000b, 00000000b, 11111111b, 11000011b, 01111110b, 00011000b, 00000000b, 00000000b

	exitOption1 db  00000000b, 00000000b, 11000011b, 00000000b, 00000000b, 11110011b, 00000000b, 00000000b
	exitOption2 db  00000000b, 00000000b, 01100110b, 00000000b, 00000000b, 11011011b, 00000000b, 00000000b
	exitOption3 db  00000000b, 00000000b, 00111100b, 00000000b, 00000000b, 11011011b, 00000000b, 00000000b
	exitOption4 db  00000000b, 00000000b, 00011000b, 00000000b, 00000000b, 11011011b, 00000000b, 00000000b
	exitOption5 db  00000000b, 00000000b, 00011000b, 00000000b, 00000000b, 11011110b, 00000000b, 00000000b
		EGMCurrOption  dw 0

	PvWhat db ? ; if its 0, its PvP, and if 1, PvAI
	gameStage db 0 ; 0 for menu, 1 for in-game, 2 for game over

CODESEG

	include AI.asm
	include convert.asm
	include graphics.asm
	include logic.asm
	include menu.asm
	include renderer.asm
	include select.asm

	start:
		mov ax, @data
		mov ds, ax

		; switch to graphics mode
			mov ax, 13h
			int 10h

		; init mouse
			mov ax, 0
			int 33h

		; show it
			mov ax, 1
			int 33h

		call startMenu

		inc byte ptr [gameStage] ; inc game stage to 1 - in-game

		call renderBoard

		mov al, true
		mainGameLoop:
			call didGameOver
			cmp al, true
				jne gameIsNotOver
				jmp gameOverScreen
			gameIsNotOver:
			call nextTurnSet
			cmp byte ptr [PvWhat], 0
				je playVSPlayer

			cmp byte ptr [turn], PLAYER_2_TURN
			 	jne playVSPlayer

			call AITurn

			jmp mainGameLoop

			playVSPlayer:
			call nextTurn
		jmp mainGameLoop

		gameOverScreen:
		inc byte ptr [gameStage] ; inc game stage to 2 - game over

    ; hide the cursor
      mov ax, 2
      int 33h

		push 200
		push 320
		push 0
		push 0
		push 0
		call drawSquare

		; game over screen
		push 15
		push offset gameOverTitle1r1
		push RMMarginLeft
		push 80
			call renderWord
		push 15
		push offset gameOverTitle2r1
		push RMMarginLeft
		push 80 + 5*3 + 5
			call renderWord

    ; bring the cursor back again
      mov ax, 1
      int 33h

		call getMouseAction
		call finishProgram


	proc nextTurnSet

		xor byte ptr [turn], 1 ; flip the turn

		; calc the color of player
			xor bh, bh
			mov bl, [player1Color]
				cmp byte ptr [turn], PLAYER_1_TURN
					je NTSAfterColorSet
				player_2_turn_color:
					mov bl,  [player2Color]
			NTSAfterColorSet:

		; hide the cursor before printing: will not print on cursor location otherwise
			mov ax, 2
			int 33h

		; render panel left
			push 200
			push 60
			push 0
			push 0
			push bx
			call drawSquare

		; render panel right
			push 200
			push 60
			push 260
			push 0
			push bx
			call drawSquare

		; bring the cursor back again
			mov ax, 1
			int 33h

		ret
	endp nextTurnSet


	; a game is over when there are no more possible moves / no more players from a specific side

	DGOP1PlayersCount equ byte ptr [bp-2]
	DGOP2PlayersCount equ byte ptr [bp-4]
	DGOP1MovesCount 		equ byte ptr [bp-6]
	DGOP2MovesCount 		equ byte ptr [bp-8]

	proc didGameOver
		push bp
		mov bp, sp
		sub sp, 8
		push bx
		push cx
		push dx

		mov DGOP1PlayersCount, 0
		mov DGOP2PlayersCount, 0
		mov DGOP1MovesCount, 0
		mov DGOP2MovesCount, 0

		; looping over the possibles moves
		mov cx, 1 ; players at the first row can't move anyway
		DGOHeightLoop:

			mov dx, 0
			DGOWidthLoop:

				push dx
				push cx
				call boardLocationToDatasegLocation

				push bx
				call getPlayer
				cmp al, NO_PLAYER_CODE
					je DGONoPlayersInc

				cmp al, PLAYER_1_CODE
					je DGOIncPlayer1

				inc DGOP2PlayersCount
				cmp cx, 0
					je DGONoPlayersInc
				inc DGOP2MovesCount
				jmp DGONoPlayersInc

				DGOIncPlayer1:
					inc DGOP1PlayersCount
					cmp cx, 7
						je DGONoPlayersInc
					inc DGOP1MovesCount

				DGONoPlayersInc:

				inc dx
			cmp dx, 8
				jb DGOWidthLoop

			inc cx
		cmp cx, 8
			jb DGOHeightLoop

		mov al, false
		cmp DGOP1PlayersCount, 0
			je DGORetrunTrue
		cmp DGOP2PlayersCount, 0
			je DGORetrunTrue
		cmp DGOP1MovesCount, 0
			je DGORetrunTrue
		cmp DGOP2MovesCount, 0
			je DGORetrunTrue

		jmp DGOFinish

		DGORetrunTrue:
			mov al, true

		DGOFinish:

		pop dx
		pop cx
		pop bx
		add sp, 8
		pop bp
		ret
	endp didGameOver

END start
